### PR TITLE
feat: notifier behind nats

### PR DESCRIPTION
**Pull request by Kishan789dev**
Fixes [#5253](https://github.com/devtron-labs/devtron/issues/5253)

Introducing "NATS," designed to facilitate the dissemination of messages published on specific topics through the implementation of the publish-subscribe model. Notifier subscribes to specified topic and starts listening

### CODE REVIEW

File: pubsub-lib/JetStreamUtil.go [hunk 1]
```diff
@@ -336,7 +341,7 @@ func AddStream(js nats.JetStreamContext, streamConfig *nats.StreamConfig, stream
 func checkConfigChangeReqd(existingConfig *nats.StreamConfig, toUpdateConfig *nats.StreamConfig) bool {
 	configChanged := false
 	newStreamSubjects := GetStreamSubjects(toUpdateConfig.Name)
-	if ((toUpdateConfig.MaxAge != time.Duration(0)) && (toUpdateConfig.MaxAge != existingConfig.MaxAge)) || (len(newStreamSubjects) != len(existingConfig.Subjects)) {
+	if ((toUpdateConfig.MaxAge != time.Duration(0)) && (toUpdateConfig.MaxAge != existingConfig.MaxAge)) || (len(newStreamSubjects) != len(existingConfig.Subjects) || (toUpdateConfig.Replicas != existingConfig.Replicas)) {
```
Comment (gireesh-devtron) [inline pubsub-lib/JetStreamUtil.go:344]: remove this replicas logic

File: pubsub-lib/NatsClient.go [hunk 2]
```diff
@@ -47,6 +47,7 @@ type NatsClientConfig struct {
 	NatsMsgBufferSize    int `env:"NATS_MSG_BUFFER_SIZE" envDefault:"-1"`
 	NatsMsgMaxAge        int `env:"NATS_MSG_MAX_AGE" envDefault:"86400"`
 	NatsMsgAckWaitInSecs int `env:"NATS_MSG_ACK_WAIT_IN_SECS" envDefault:"120"`
+	Replicas             int `env:"REPLICAS" envDefault:"0"`
```
Comment (gireesh-devtron) [inline pubsub-lib/NatsClient.go:50]: remove it

File: pubsub-lib/NatsClient.go [hunk 3]
```diff
@@ -62,6 +63,7 @@ func (ncc NatsClientConfig) GetDefaultNatsConsumerConfig() NatsConsumerConfig {
 		NatsMsgProcessingBatchSize: ncc.NatsMsgProcessingBatchSize,
 		NatsMsgBufferSize:          ncc.GetNatsMsgBufferSize(),
 		AckWaitInSecs:              ncc.NatsMsgAckWaitInSecs,
+		Replicas:                   ncc.Replicas,
```
Comment (gireesh-devtron) [inline pubsub-lib/NatsClient.go:66]: remove it

File: pubsub-lib/NatsClient.go [hunk 4]
```diff
@@ -74,8 +76,10 @@ func (ncc NatsClientConfig) GetDefaultNatsStreamConfig() NatsStreamConfig {
 }
 
 type StreamConfig struct {
-	MaxAge time.Duration `json:"max_age"`
+	MaxAge   time.Duration `json:"max_age"`
+	Replicas int           `json:"num_replicas"`
```
Comment (gireesh-devtron) [inline pubsub-lib/NatsClient.go:80]: remove it

File: pubsub-lib/NatsClient.go [hunk 5]
```diff
@@ -88,6 +92,8 @@ type NatsConsumerConfig struct {
 	NatsMsgBufferSize int `json:"natsMsgBufferSize"`
 	// AckWaitInSecs is the time in seconds for which the message can be in unacknowledged state
 	AckWaitInSecs int `json:"ackWaitInSecs"`
+	//it will show the instances created for the consumers on a particular subject(topic)
+	Replicas int `json:"replicas"`
```
Comment (gireesh-devtron) [inline pubsub-lib/NatsClient.go:96]: remove it

File: pubsub-lib/PubSubClientService.go [hunk 6]
```diff
@@ -27,6 +27,7 @@ type LoggerFunc func(msg model.PubSubMsg) (logMsg string, keysAndValues []interf
 type PubSubClientService interface {
 	Publish(topic string, msg string) error
 	Subscribe(topic string, callback func(msg *model.PubSubMsg), loggerFunc LoggerFunc, validations ...ValidateMsg) error
+	isClustered()
```
Comment (gireesh-devtron) [inline pubsub-lib/PubSubClientService.go:30]: remove this method, why did we add this function if it's not exported

File: pubsub-lib/PubSubClientService.go [hunk 7]
```diff
@@ -136,6 +138,8 @@ func (impl PubSubClientServiceImpl) Subscribe(topic string, callback func(msg *m
 		return err
 	}
 	go impl.startListeningForEvents(processingBatchSize, channel, callback, loggerFunc, validations...)
+
+	//time.Sleep(10 * time.Second)
```
Comment (gireesh-devtron) [inline pubsub-lib/PubSubClientService.go:142]: remove this comment

File: pubsub-lib/PubSubClientService.go [hunk 8]
```diff
@@ -242,6 +247,12 @@ func (impl PubSubClientServiceImpl) TryCatchCallBack(msg *nats.Msg, callback fun
 	callback(subMsg)
 }
 
+func (impl PubSubClientServiceImpl) isClustered(streamName string) bool {
+	// This is only ever set, no need for lock here.
+	streamInfo, _ := impl.NatsClient.JetStrCtxt.StreamInfo(streamName)
+	return streamInfo.Cluster != nil
+}
+
```
Comment (gireesh-devtron) [inline pubsub-lib/PubSubClientService.go:255]: remove this method

File: pubsub-lib/PubSubClientService.go [hunk 9]
```diff
@@ -282,6 +293,17 @@ func (impl PubSubClientServiceImpl) updateConsumer(natsClient *NatsClient, strea
...
+		if replicas > 1 && impl.isClustered(streamName) {
+			existingConfig.Replicas = replicas
+			updatesDetected = true
+		} else {
+			if replicas > 1 {
+				impl.Logger.Errorw("replicas >1 is not possible in non-clustered mode ")
+			}
+		}
+
+	}
```
Comment (gireesh-devtron) [inline pubsub-lib/PubSubClientService.go:306]: remove this

**Review by kripanshdevtron — APPROVED** [review]

**Pull request merged**